spanner/api/Program.cs

```diff
@@ -0,0 +1,965 @@
+            // Create client.
+            DatabaseAdminClient databaseAdminClient =
+                await DatabaseAdminClient.CreateAsync();
+            // Initialize request argument(s).
+            InstanceName parent = new InstanceName(projectId, instanceId);
+            string createStatement =
+                $"CREATE INDEX AlbumsByAlbumTitle2 ON Albums(AlbumTitle) "
+                + "STORING (MarketingBudget)";
+            // Make the request.
+            var createDatabaseRequest =
+                await databaseAdminClient.UpdateDatabaseDdlAsync(
+                    new DatabaseName(projectId, instanceId, databaseId),
+                new[] { createStatement });
+            // Poll until the returned long-running operation is complete.
+            var updateDatabaseResponse =
+                await createDatabaseRequest.PollUntilCompletedAsync();
+            return updateDatabaseResponse.Result;
+            // [END create_storing_index]
+        }
+
+        public static async Task<object> AddColumnAsync(
+    string projectId, string instanceId, string databaseId)
+        {
+            // [START add_column]
+            // Create client.
+            DatabaseAdminClient databaseAdminClient =
+                await DatabaseAdminClient.CreateAsync();
+            // Initialize request argument(s).
+            InstanceName parent = new InstanceName(projectId, instanceId);
+            string alterStatement =
+                $"ALTER TABLE Albums ADD COLUMN MarketingBudget INT64";
+            // Make the request.
+            var createDatabaseRequest =
+                await databaseAdminClient.UpdateDatabaseDdlAsync(
+                    new DatabaseName(projectId, instanceId, databaseId),
+                new[] { alterStatement });
+            // Poll until the returned long-running operation is complete.
+            var updateDatabaseResponse =
+                await createDatabaseRequest.PollUntilCompletedAsync();
+            Console.WriteLine("Added the MarketingBudget column.");
+            return updateDatabaseResponse.Result;
+            // [END add_column]
+        }
+
+        public static async Task QuerySampleDataAsync(
+            string projectId, string instanceId, string databaseId)
+        {
+            // [START query_data]
+            string connectionString =
+            $"Data Source=projects/{projectId}/instances/"
+            + $"{instanceId}/databases/{databaseId}";
+            // Create connection to spanner.
+            using (var connection = new SpannerConnection(connectionString))
+            {
+                var cmd = connection.CreateSelectCommand(
+                    "SELECT SingerId, AlbumId, AlbumTitle FROM Albums");
+                using (var reader = await cmd.ExecuteReaderAsync())
+                {
+                    while (await reader.ReadAsync())
+                    {
+                        Console.WriteLine($"SingerId : {reader["SingerId"]} "

```

Comment:
we had a lot of issues in firebase because users did not specify in code what type they were expecting and had to determine the magic underlying type the client library decided to return.
This isn't about what the server does.  This is about having the user use a method that declares what type he is expecting versus just casting it which is not as robust.  It also requires the value to be boxed, which is a performance penalty.

Also SqlDataReader provides this very method, so its a good idea from a competitive point of view.

So for example.

Do NOT do this:    double val = (double)reader["SomeColumn"];  <== boxes value, will easily break on schema changes.

DO this:  double val = reader.GetFieldValue{double}("SpannerColumn");  <== no boxing, more robust conversion in the client library.

we will do conversions for several types.  For example, complex types like structs and lists, you can pass in a concrete custom type derived from generics IList{T} and IDictionary{T}.
We'll also do date conversions to an underlying string format.

Most this work is actually done by the .net frameworks Convert.ToX().  We just enumerate thru all the various permutations so you need to think less about the types you are getting.